pkg/apis/core/validation/validation.go

```diff
@@ -6219,33 +6236,129 @@ func validatePodResizeContainerOrdering(newPod, oldPod *core.Pod, specPath *fiel
+	var allErrs field.ErrorList
+	if !isPodLevelResourcesResizeRequestSupported(*oldPod) {
+		allErrs = append(allErrs, field.Forbidden(specPath, fmt.Sprintf("Pod running on node without support for pod-level resources resize %v", oldPod.Status.Resources)))
+	}
+
+	if oldPod.Spec.Resources != nil && newPod.Spec.Resources == nil {
+		return append(allErrs, field.Forbidden(specPath.Child("resources"), "pod-level resources cannot be removed"))
+	}
+
+	if oldPod.Spec.Resources != nil {
+		if oldPod.Spec.Resources.Requests != nil {
+			if newPod.Spec.Resources.Requests == nil || resourcesRemoved(newPod.Spec.Resources.Requests, oldPod.Spec.Resources.Requests) {
+				allErrs = append(allErrs, field.Forbidden(specPath.Child("resources").Child("requests"), "pod-level resource requests cannot be removed"))
+			}
 		}
-		if mem, found := oldResourceList[core.ResourceMemory]; found {
-			mungedResourceList[core.ResourceMemory] = mem
+
+		if oldPod.Spec.Resources.Limits != nil {
+			if newPod.Spec.Resources.Limits == nil || resourcesRemoved(newPod.Spec.Resources.Limits, oldPod.Spec.Resources.Limits) {
+				allErrs = append(allErrs, field.Forbidden(specPath.Child("resources").Child("limits"), "pod-level resource limits cannot be removed"))
+			}
 		}
-		return mungedResourceList
 	}
+
+	dropCPUMemoryResourcesFromPodLevelResources(podSpecToMutate.Resources, oldPod.Spec.Resources)
+
+	// Drop Resources from the new PodSpec if the old PodSpec had no resources set
+	// and the new PodSpec does not include the immutable 'Resources.Claims' field.
+	if oldPod.Spec.Resources == nil && len(podSpecToMutate.Resources.Claims) == 0 {
+		podSpecToMutate.Resources = nil
+	}
+
+	if !apiequality.Semantic.DeepEqual(podSpecToMutate.Resources, oldPod.Spec.Resources) {
+		// This likely means that the user has made changes to pod-level resources other
+		// than CPU and memory.
+		errs := field.Forbidden(specPath, "only cpu and memory resources are mutable at pod-level")
+		allErrs = append(allErrs, errs)
+	}
+
+	return allErrs
+}
+
+func dropCPUMemoryUpdates(resourceList, oldResourceList core.ResourceList) core.ResourceList {
+	if oldResourceList == nil {
+		return nil
+	}
+	var mungedResourceList core.ResourceList
+	if resourceList == nil {
+		mungedResourceList = make(core.ResourceList)
+	} else {
+		mungedResourceList = resourceList.DeepCopy()
+	}
+	delete(mungedResourceList, core.ResourceCPU)
+	delete(mungedResourceList, core.ResourceMemory)
+	if cpu, found := oldResourceList[core.ResourceCPU]; found {
+		mungedResourceList[core.ResourceCPU] = cpu
+	}
+	if mem, found := oldResourceList[core.ResourceMemory]; found {
+		mungedResourceList[core.ResourceMemory] = mem
+	}
+	return mungedResourceList
+}
+
+// dropCPUMemoryResourcesFromContainer deletes the cpu and memory resources from the container, and copies them from the old pod container resources if present.
+func dropCPUMemoryResourcesFromContainer(container *core.Container, oldPodSpecContainer *core.Container) {
```

Comment:
nit: delete this function: just use `dropCPUMemoryResourcesFromPodLevelResources` (renamed) for containers too. E.g. `dropCPUMemoryResources(&container.Resources, &oldContainer.Resources)`

Reply:
actually, because of the "Set the entire Resources block to nil if two conditions are met" block in that method, let's defer collapsing to a single method to a follow up